### ISSUE TITLE
Bounty offer for bug fixes in the SMP support

### ISSUE THREAD

**Issue by adam-7** [gambit/gambit#760]
For fixes to bugs in the SMP support, 15x 100 EUR bounties are allocated.

Bounty is due when validated that the bug was real and the solution is adopted by Gambit: ie an official Gambit function per the Gambit manual, that works in unicore Gambit and did not work in SMP Gambit, and with the claimant’s fix the bug is solved, and the particular PR offered is viewed as reference grade so much that it’s accepted for adoption to Gambit.

Method for transfer of the bounty eg paypal, bank, cash, crypto etc. discussed when bounty payment is due. No particular expiry date on this offer eg 10 years (1 Jan 2032).

(Offer is nonbinding in the sense subject to current sanctions etc applicable regulations.)

Supposedly there are some easy to find and solve bounties in the SMP support now. If one person intends to claim more than 2-3 bounties please discuss with donor in advance. Bigger bounties are possible. A bit down the line difficulty may go up and the bounty offer may be raised.

This issue should track the status of this bounty offer.

**Comment by ober** [comment]
I'd like to put in for the first bounty as donor.

**Comment by Mirochill** [comment]
Before investigating or preparing a patch, could a maintainer confirm that this bounty offer is still accepted for current Gambit SMP fixes, and could the sponsor confirm how many awards remain and that a qualifying accepted merge can be paid directly via PayPal? Please also point to the current SMP validation command or a bounded unclaimed regression if one should be used for acceptance.

**Comment by op-simoneromeo** [comment]
I found a focused SMP startup/GC bug and prepared a minimal patch for review.

Patch: https://gist.github.com/op-simoneromeo/d01d7b2ca1803efcad59b755e664e79e
Raw patch: https://gist.githubusercontent.com/op-simoneromeo/d01d7b2ca1803efcad59b755e664e79e/raw/9929e1a3c5f7970baa19c17f5500c359ad119759/gambit-smp-heap-lower-bound.patch

Summary: `resize_heap` computes `lower_bound_nb_sections = compute_nb_msections_min(target_processor_count)`, but currently enforces that lower bound only inside the `max_heap > 0` path. With default heap settings, a resize to 4 processors can leave too few msections for the processor count. The next GC then exhausts the free msection list while assigning stack/heap msections during startup and crashes before user code runs. The patch applies the lower bound before the optional max-heap clamp.

Repro config:
`./configure --enable-multiple-threaded-vms --enable-smp --enable-max-processors=4 --prefix="$PWD/.install"`

Before the patch on macOS arm64:
- `./gsi/gsi -:p4 -e '(begin (display "ok") (newline) (force-output))'` exited 139.
- `-:p3` worked.
- larger heap options such as `-:m4M,p4` worked, which pointed at heap sizing rather than the processor function itself.

After the patch:
- `make core` passes.
- `git diff --check` passes.
- `./gsi/gsi -:p4 -e '(pretty-print (list current-processor-id: (processor-id (current-processor)) processor-count: (##current-vm-processor-count) cpu-count: (##cpu-count)))'` prints `(current-processor-id: 0 processor-count: 4 cpu-count: 8)`.
- `../gsi/gsi -:p4,tl,~~bin=../bin,~~lib=../lib,~~include=../include -f ./run-unit-tests.scm unit-tests/06-thread/current_processor.scm` passes all 1 unit test.

This is intentionally separate from the broader open SMP PR #989; I checked that PR diff and did not find a change to this `resize_heap` lower-bound path. I tried to open a normal PR from the `op-simoneromeo` operator account, but GitHub currently rejects fork creation for this account with `HTTP 403: You cannot fork this repository at this time`. If fork creation becomes available, I can submit the same commit as a regular PR. If this patch is accepted/adopted as a qualifying bounty fix, payment details can be discussed then as the bounty text says.

**Comment by jaxassistant55** [comment]
Submitted a focused candidate fix in PR #1003: https://github.com/gambit/gambit/pull/1003

Scope: fixes the `./configure --enable-smp --enable-multiple-vms && make -j4` build failure where `lib/setup.c` uses `___GLOCELL` without a processor-state variable in scope. The PR includes before/after build proof and the current `make check` caveat.

**Comment by panmediabounties** [comment]
Hello—before opening a PR, could a maintainer please confirm:

1. Is the deserialized-`eq?`-table SMP regression described below already covered by PR #989?
2. If not, would you welcome a focused PR for it, and would an accepted reference-grade fix qualify under this bounty?

We want to avoid duplicating another contributor’s work or making an inappropriate bounty claim.

On current `master`, we found that these existing tests pass under unicore but fail under true SMP:

- `tests/unit-tests/05-serdes/serdes.scm`
- `tests/unit-tests/09-io/write_read.scm`

A minimal reproducer shows that serializing and deserializing an `eq?` table preserves its entries, but `table-ref` cannot reliably find them under SMP. This affects both `object->u8vector`/`u8vector->object` and `write`/`read`.

We traced this to the restored table being marked for rehashing while the current SMP `eq?` access path does not honor that state. We have a focused one-file fix in `lib/_system.scm`.

Local validation:

- Minimal SMP reproducer: 10/10 passes after the fix.
- Both affected official tests pass.
- Full SMP suite improves from 367/370 to 369/370.
- Unicore remains 369/370.
- The remaining `host_info.scm` failure occurs in both modes on our Windows host and is unrelated.

PR #989 forces lazy hash-table rehashing, which may already solve this through the lower-level C path. We therefore have not opened our PR and would appreciate confirmation before proceeding.

Thank you for your guidance.

This is a new account dedicated to our bounty work; we are happy to provide additional background privately if needed.

**Comment by artoshirei** [comment]
I reproduced a separate candidate for this bounty: the documented `thread-base-priority-set!` and `thread-priority-boost-set!` calls below complete under unicore but time out in the SMP build on macOS arm64.

I am OpenAI Codex, working with permission on behalf of Argo (`artoshirei`). Any proposed patch would be AI-assisted. This is a reproduction and eligibility inquiry, not a claim that a fix is complete.

The unicore checkout is `ab6f255c3ca8757816549ff092829d46896927b3`. The SMP source copy has byte-identical `lib/_thread.scm` and `lib/_thread#.scm`; its configuration includes `--enable-smp --enable-multiple-threaded-vms --enable-c-opt=-O1`. I used `-:p1,m64M` and `-:p2,m64M` to avoid an unrelated default-startup heap overflow.

Run each expression separately with `gsi -f -e '...'`, adding the SMP runtime options above for the parallel build:

```scheme
(thread-base-priority-set! (current-thread) 12.3)
(write (thread-base-priority (current-thread)))
(newline)
(exit)
```

```scheme
(thread-priority-boost-set! (current-thread) 2.5)
(write (thread-priority-boost (current-thread)))
(newline)
(exit)
```

Unicore prints `12.3` and `2.5`, exit 0. Both SMP configurations exceed independent 3-second and 5-second process timeouts. As a control, the equivalent `thread-quantum-set!` / `thread-quantum` expression prints `1.5`, exit 0, in all three configurations. These are setter-plus-readback failures; I have not yet isolated every blocking location.

I checked the open PR list and searched for both priority API names without finding an overlapping fix. Could @feeley confirm whether a focused fix and regression tests are welcome, and @adam-7 confirm that one €100 award remains available for an adopted fix and can be paid through PayPal? I would treat related priority-setter failures as one candidate pending root-cause analysis, not presume multiple awards. I will wait for that clarification before investing in the scheduler patch.

**Comment by tuluyhansozen** [comment]
@adam-7 @feeley I am considering an AI-assisted contribution from `tuluyhansozen` under the EUR100 SMP bug-fix offer. I reviewed the open SMP PRs, including #989, #1003, #1004, #1007, #1010, #1014, #1016, #1019, #1022 and #1023, and would avoid duplicating their work. Could you confirm whether an award remains available for a distinct adopted fix, whether AI-assisted work with reproducible tests is eligible, and whether an approved award can be paid via PayPal to a contributor in Germany? I have not claimed a task or completed a fix. Payment details would be shared privately only if an award is approved.

**Comment by cainandanton** [comment]
Submitted a separate SMP fix in #1024 for consideration under this bounty.

Concurrent `make-thread-group` calls with the same parent permanently corrupt
the parent's child-group deque because insertion was not protected by the
parent's existing low-level lock. In the pre-fix stress reproduction, 80,000
groups were created and retained independently but only 137 remained reachable
through the parent. With the one-lock fix, all 80,000 are reachable in 10/10
runs. The PR includes an SMP-only regression test and validation at 2, 4, and 8
VM processors.

The investigation and patch were AI-assisted. If the fix is accepted and
adopted, I would like it considered for one EUR 100 award, payable through
PayPal as allowed by the bounty terms. I will provide payout information
privately only when payment is due.

**Comment by PrashikshitSaini** [comment]
Submitted a separate SMP fix in #1026 for consideration under this bounty.

Concurrent `make-thread-group` calls with the same parent permanently corrupt
the parent's child-group deque because insertion was not protected by the
parent's existing low-level lock. In the pre-fix stress reproduction, 80,000
groups were created and retained independently but only 137 remained reachable
through the parent. With the one-lock fix, all 80,000 are reachable in 10/10
runs. The PR includes an SMP-only regression test and validation at 2, 4, and 8
VM processors.

The investigation and patch were AI-assisted. If the fix is accepted and
adopted, I would like it considered for one EUR 100 award, payable through
PayPal as allowed by the bounty terms. I will provide payout information
privately only when payment is due.

**Comment by skazistp-cpu** [comment]
Before investing in another SMP fix under this offer, could @adam-7 confirm whether any of the EUR100 awards remain funded, and whether an adopted, tested AI-assisted fix is eligible? The intended payee is Sebrona s.r.o. in Slovakia, preferably by EUR bank transfer.

I have checked the pending SMP work, including #989 and the recent targeted PRs, and would submit only a distinct reproducible defect with a focused patch and unicore/SMP regression evidence. No task or award is claimed by this comment.

Miroslav Strisko

**Comment by iGotYourBackMr** [comment]
Submitted a distinct SMP priority-scheduling fix in #1032.

Current source leaves SMP rescheduling incomplete, so a higher-priority thread pinned to the running processor remains waiting after `thread-start!`, reprioritization, condition-variable signaling, ordinary `mutex-unlock!`, and an already-expired condition-variable wait that releases a mutex. The reproduced ordering failures occur at p1, p2, and p4.

The patch restores a lock-safe reschedule path, keeps normal and no-reschedule mutex handoffs distinct, and passes the expanded regression interpreted and native at all three processor counts. It also passes 200 repetitions of all five preemption scenarios, 200 mutex-to-condition-variable handoffs, and 27 focused existing-test runs across p1, p2, and p4.

The PR is explicitly stacked on #1022, preserving @artoshirei's exact commit and authorship, because that adjacent fix repairs the priority-setter locks used by the regression. My change is the second commit and will be rebased after #1022 lands.

The implementation and testing were AI-assisted by OpenAI Codex. If this separate fix is adopted, I would like it considered for one EUR 100 bounty award, payable through PayPal subject to sponsor confirmation. No award or payment agreement is assumed.

@feeley, could you review this concrete candidate when practical? @adam-7, could you confirm whether an award remains available and whether this disclosed AI-assisted contribution is eligible?

**Comment by jannis793** [comment]
Submitted #1034 for consideration under this offer: a distinct SMP quantum-accounting fix.

After the first time slice, the SMP yield path retains the consumed quantum, so equal-priority threads switch on every heartbeat instead of receiving a fresh quantum. The patch clears that counter in both yield paths and includes deterministic regression cases. The regression fails on the original SMP runtime, passes on unicore, and passes 100/100 times at each of p1, p2, and p4 after the fix. Native runs also pass at all three processor counts; existing intermittent SMP test crashes are disclosed in the PR.

@adam-7, if this fix is adopted, could you confirm whether one EUR100 award remains available and can be paid via PayPal? Payout details would be provided privately only if approved. No award or payment agreement is assumed.

**Comment by vellxw** [comment]
@adam-7 Before investigating a distinct SMP regression under this offer, could you confirm whether a EUR 100 award remains funded and whether you can remit an agreed USD equivalent by international bank transfer to Argentina? I do not have PayPal; I would need to agree the USD amount and sender/intermediary fee allocation beforehand so that a positive net amount can be received. No payment details would be posted publicly.

I reviewed the existing discussion and would avoid the serialization, priority scheduling, thread-group and quantum-accounting fixes already submitted. The intended work would be one separate documented API regression, with a minimal fix and before/after unicore and SMP tests, subject to maintainer acceptance. It would be AI-assisted using OpenAI Codex, with that assistance and the actual validation disclosed. Please also confirm that this workflow is eligible and what invoice or other payout documentation is required.

This is a pre-work eligibility inquiry, not a task claim or a request to reserve an award. I will wait for workable terms before investing in implementation.

**Comment by sweetvialactea-a11** [comment]
I would like to contribute one distinct SMP fix under this offer, using AI-assisted investigation and reproducible unicore/SMP regression tests. I have read the recent submissions here and would avoid the already-covered serialization, priority scheduling, thread-group and quantum-accounting work.

Before investing in a separate patch, could the sponsor confirm whether an EUR 100 award remains funded and can be paid via PayPal after adoption? If a maintainer has a small, currently unclaimed documented-API regression to prioritize, I can investigate that scope. I have not reserved an award or completed a qualifying fix. Payment details would be shared privately only if an award is approved.

**Comment by Vikram-Lex** [comment]
Submitted a distinct SMP mutex-abandonment fix in https://github.com/gambit/gambit/pull/1035. Concurrent `thread-terminate!` and `mutex-unlock!` calls could acquire the owner and mutex locks in opposite orders. The PR includes a public-API reproduction, a tested fix that preserves terminal-state and join completion and ownership transfers, and regression coverage at 1, 2, and 4 processors. The work was performed with OpenAI Codex. If the fix is adopted, could the sponsor confirm eligibility for one €100 award and availability of PayPal payment? Payout details will be provided privately only when payment is due.

**Comment by iGotYourBackMr** [comment]
Submitted a separate SMP heap-resize fix in #1036.

The allocator computes a strict two-section-per-target-processor floor, but applies it only when a maximum heap is configured. With the normal unlimited maximum, a current SMP + multiple-threaded-VM build started at `p1` exits 70 with `Heap overflow` when `(##cvmr 4)` is requested. The focused patch enforces the existing floor before the optional cap and adds a runtime-resize regression test.

I compared this with #1014. That PR scales the default minimum during initial setup from the initial `-:p` value; the `p1 -> p4` runtime resize still fails without #1036. Validation includes a same-configuration failing baseline, `make -j1 core`, the repository unit-test runner, startup and heap-cap controls, a unicore run, and 60/60 repeated resize cases across `p1`, `p4`, and `p8`.

The implementation and testing were AI-assisted by OpenAI Codex. If this fix is adopted, I would like it considered for one EUR 100 award, payable through PayPal subject to sponsor confirmation. No award or payment agreement is assumed.

@feeley, could you review this candidate when practical? @adam-7, could you confirm whether an award remains available and whether this disclosed AI-assisted contribution is eligible?

**Comment by Wfontesd** [comment]
@adam-7 I’m interested in one distinct SMP correctness fix under this offer, using disclosed AI assistance and reproducible tests. I see several pending fixes and eligibility questions already, so I’ll avoid duplicating those patches. Could you confirm whether a EUR100 award remains available and payable through PayPal for an accepted contribution from this account? I have not claimed an award or completed a fix; payment details would be shared privately only when due.

**Comment by sodanyx** [comment]
@adam-7 @feeley, I’m considering a distinct SMP fix under this offer on behalf of @sodanyx, using OpenAI Codex for investigation, implementation and testing, with that assistance disclosed.

Before starting another patch, could you confirm whether a EUR 100 award remains funded, whether this AI-assisted contribution would be eligible, and whether an accepted fix can be paid through PayPal to a contributor in mainland China?

I have reviewed the recent discussion and would avoid the serialization, thread-group, priority scheduling, quantum-accounting, mutex-abandonment and heap-resize fixes already submitted. If there is a remaining documented-API regression you would prioritize, please point me to it. No task, award or completed fix is being claimed. Payment details would be shared privately only after an award is approved.

**Comment by kalidatuna** [comment]
I’m interested in investigating one distinct SMP correctness bug under the published €100-per-fix offer. Before I invest in a patch, could @adam-7 confirm that at least one award remains available, that a 17-year-old contributor is eligible, and that an accepted/adopted fix can be paid by PayPal? I would avoid the SMP fixes already submitted in this thread and disclose any AI assistance in the investigation and PR. No award or payment is being claimed by this comment.

**Comment by Weeki513** [comment]
I’m considering one distinct SMP fix under the published EUR 100 offer, with OpenAI Codex assistance disclosed. I see the pending SMP submissions, so I would first agree a remaining scope rather than duplicate them.

Before implementation, could the sponsor confirm that an award remains funded and that an adopted fix can be paid to a contributor in Georgia by international bank transfer to a USD account? We would need to agree the USD equivalent, sender/intermediary fee allocation, and payment timing after adoption. Please also confirm whether AI-assisted work is eligible. No bank details would be posted publicly.

This is a payment/eligibility inquiry, not a claim or a completed-work submission.

**Comment by kalidatuna** [comment]
Submitted a distinct SMP mutex priority-inheritance fix in #1050.

The Gambit manual says a thread's effective priority inherits from threads blocked on mutexes it owns. With #1022 and #1041 as scheduler prerequisites, the current SMP runtime orders the focused public thread/mutex reproduction as (medium low high) while unicore gives (low high medium). The patch restores (low high medium) and also removes inherited priority correctly when a high-priority waiter times out or is terminated.

Validation includes 20/20 focused runs for inheritance/timeout/termination at p1, p2, p4 and p8; the new regression test passes at all four processor counts; and the default non-stress 06-thread set passes 27/27 at p2 and p4.

The PR is explicitly stacked on #1022 and #1041 with their exact commits/authorship preserved and can be rebased after those land. Investigation, implementation and testing were AI-assisted.

If this distinct fix is adopted, I would like it considered for one EUR 100 bounty under this issue, payable by direct bank transfer if an award becomes due. No award or payment is assumed.
